### DISCUSSION TITLE
Ability to add Date Made filter for cookbooks.

### DISCUSSION THREAD

**Discussion by ghost** [mealie-recipes/mealie#4729] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

When creating cookbooks, there doens't appear to be a way to filter recipes based on their "Last Made" date. This would be helpful so that you can set up a cookbook that shows what you've cooked in some arbitrary timeframe (last week, this year, etc).

### Please provide a concise description of the feature that would resolve your issue.

I would like the ability to create a cookbook that shows all recipes that have been added but not cooked yet. I'm importing family recipes and would like to keep tabs on which ones I haven't tried yet, and pick from the list.

### Please consider and list out some caveats or tradeoffs made in your design decision

I cannot think of any tradeoffs or pitfalls.

### Additional Information

- [ ] If this is accepted I'm willing to submit a PR to provide this feature
- [ ] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by ChipSkylark37** [comment]
I've been thinking of a way to do the same. For now, I add a "tag" to the recipes I haven't tried as a generic "want to try" tag. Then I can filter that when making a recipe. 

It's not perfect, and I'd like a button that just says "I want to make this" so when I'm searching for a recipe to make, there's a "try something you haven't made before" which is sort of like the "roll the dice" button.

**Comment by Ionshard** [comment]
DISCLAIMER: Today is the first that I started looking at the mealie codebase. Everything I say from here on out should be treated as the rantings of someone who has no clue what they're talking about!

Anyways, I ended up in this discussion and I also wanted this feature. I found it odd that I could sort by Last Made on the recipe page but there was no filter in the cookbook page. So I started reading the code to see what I could find and I saw that the cookbooks simply store a `query_filter_string` that is used to filter the recipes when showing the cookbook.

So I (made a backup and then) modified my `mealie.db` to change my "To Try" cookbook and manually set the `query_filter_string` to be `'last_made is NULL'` and the cookbook starts showing me all recipes without a last made date! So it's definitely possible to get the cookbooks to filter based off the last_made column! So now I have my cookbook working as desired, though I have absolutely no clue if this will have further repercussions on the system.

I have been hopping around the code reading the various parts of the system figuring out how the server and the client works. I have never done Vue but it seems close enough to React that I could pretend I knew what I was looking at.

Thinking about how this could be implemented, Since all the `query_string_field` generation seems to be done client side, I think it would be pretty trivial to add a new `presence` type to the `FieldTypes` in [use-query-filter-builder](https://github.com/mealie-recipes/mealie/blob/mealie-next/frontend/composables/use-query-filter-builder.ts) that would check if a column is set or not. This would only use the `relOps.value["IS"]` operator (which seems to be implemented but not used by any of the currently accessible filters 🤔).

Then in the [QueryFilterBuilder](https://github.com/mealie-recipes/mealie/blob/mealie-next/frontend/components/Domain/QueryFilterBuilder.vue) it should be relatively easy to copy the boolean handler to change it for our purposes ... maybe something like:
```vue
              <v-checkbox
                v-else-if="field.type === 'presence'"
                :model-value="field.value"
                @update:model-value="setFieldValue(field, index, $event! ? 'NOT NULL' : 'NULL')"
              />
```
Yes, this is a little hacky because it puts `NOT NULL`/`NULL` in the value of the field but the `NOT` is technically part of the operator but in the end it just becomes `last_made IS NOT NULL` or `last_made IS NULL` which is a valid query string.

Finally then I guess we'd just add a new entry to the `fieldDefs` in [CookbookEditor](https://github.com/mealie-recipes/mealie/blob/mealie-next/frontend/components/Domain/Cookbook/CookbookEditor.vue) like this:
```ts
      {
        name: "last_made",
        label: i18n.t("general.last-made"),
        type: "presence",
      },
```

and I think maybe that should be enough to allow us to determine if last_made is set or not ...

Now I have no clue how you'd handle ranges, like the ability to show anything last made greater than 1 year ago and I am running out of steam to think about this currently, so that will have to be an exercise for the future.

Unfortunately thinking of "what to do" is the easy part, actually doing it is another thing ... I don't know if I have enough willpower to clone this repo, get my developer environment up and running and actually play around with this to test my theory ... I dunno maybe I'll get around to this in the future, but since my hack in the database does exactly what I want my motivation is a little low. Especially if simple "presence" detection in the query builder isn't good enough and this might all get rejected in PR anyways.

So hopefully maybe this message is enough for someone else to take the reigns and give this a go!

**Comment by nununo** [comment]
I too would like this. I would like to be able to have a cookbook listing all the recipes that I never tried.

**Comment by bpmartin20** [comment]
I agree with the desire. I'd like to be able to include as a recipe criteria "last-made > x (days|weeks|months)" so I don't select a recipe we've just had recently.

Similarly, I'd like to be able to include the same kind of treatment for recipe ratings. In our house, 3 stars means "suitable for company", so I'd like a cookbook  that selects "category=Main Courses and rating >= 3 and tagged Gluten-free" for example.

An "advanced user" ability to write and save our own queries would be really cool, if you documented the commonly used field names, but maybe that's too much for now.

By the way, the above information about internal field names appears to be out of date. I've just examined my mealie.db database (SQLIte), and there is no query_filter_string field in the cookbooks table or anywhere else. I'm sure there's something like it somewhere, but not in cookbooks and not by that name.
